be/src/vec/core/sort_cursor.h

```diff
@@ -21,43 +21,123 @@
-#include "vec/runtime/vdata_stream_recvr.h"
 
 namespace doris::vectorized {
 
+struct HeapSortCursorBlockView {
+public:
+    Block block;
+    ColumnRawPtrs sort_columns;
+    SortDescription& desc;
+
+    HeapSortCursorBlockView(Block&& cur_block, SortDescription& sort_desc)
+            : block(cur_block), desc(sort_desc) {
+        _reset();
+    }
+
+    void filter_block(IColumn::Filter& filter) {
+        Block::filter_block_internal(&block, filter, block.columns());
+        _reset();
+    }
+
+private:
+    void _reset() {
+        sort_columns.clear();
+        auto columns = block.get_columns();
+        for (size_t j = 0, size = desc.size(); j < size; ++j) {
+            auto& column_desc = desc[j];
+            size_t column_number = !column_desc.column_name.empty()
+                                           ? block.get_position_by_name(column_desc.column_name)
+                                           : column_desc.column_number;
+            sort_columns.push_back(columns[column_number].get());
+        }
+    }
+};
+
+struct HeapSortCursorImpl {
+public:
+    HeapSortCursorImpl() : _row_id(0), _block_view(nullptr) {}
+    HeapSortCursorImpl(int row_id, std::shared_ptr<HeapSortCursorBlockView> block_view)
+            : _row_id(row_id), _block_view(block_view) {}
+
+    HeapSortCursorImpl(const HeapSortCursorImpl& other) {
+        _row_id = other._row_id;
+        _block_view = other._block_view;
+    }
+
+    HeapSortCursorImpl(HeapSortCursorImpl&& other) {
+        _row_id = other._row_id;
+        _block_view = std::move(other._block_view);
+    }
+
+    HeapSortCursorImpl& operator=(HeapSortCursorImpl&& other) {
+        std::swap(_row_id, other._row_id);
+        std::swap(_block_view, other._block_view);
+        return *this;
+    }
+
+    ~HeapSortCursorImpl() = default;
+    const size_t row_id() const { return _row_id; }
+
+    const ColumnRawPtrs& sort_columns() const { return _block_view->sort_columns; }
+
+    const Block* block() const { return &_block_view->block; }
+
+    const SortDescription& sort_desc() const { return _block_view->desc; }
+
+    bool operator<(const HeapSortCursorImpl& rhs) const {
+        for (size_t i = 0; i < sort_desc().size(); ++i) {
+            int direction = sort_desc()[i].direction;
+            int nulls_direction = sort_desc()[i].nulls_direction;
+            int res = direction * sort_columns()[i]->compare_at(row_id(), rhs.row_id(),
+                                                                *(rhs.sort_columns()[i]),
+                                                                nulls_direction);
+            // ASC: direction == 1. If bigger, res > 0. So we return true.
+            if (res < 0) {
+                return true;
+            }
+            if (res > 0) {
+                return false;
+            }
+        }
+        return false;
+    }
+
+private:
+    size_t _row_id;
+    std::shared_ptr<HeapSortCursorBlockView> _block_view;
```

Comment:
Smart pointers introduce additional memory and performance overhead, consider whether we can circumvent its use with other mechanisms to gain performance gains